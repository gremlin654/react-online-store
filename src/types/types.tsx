export interface IPhone {
  id: string
  model: string
  img: string
  shape: string
  quantity: number
  release: number
  color: string
  cameras: string
  popular: boolean
  favorite: string
  inCart: boolean
}

export interface IPhoneListProps {
  phonesData: IPhone[]
}

export enum LOCAL_STORAGE_KEYS {
  filterShape = 'filterShape',
  filterCameras = 'filterCameras',
  filterColor = 'filterColor',
  filterPopular = 'filterPopular',
  filterRelease = 'filterRelease',
  filterQuantity = 'filterQuantity',
  search = 'search',
  sortedByParameter = 'sortedByParameter',
}

export interface IPropsPhones {
  phones: IPhone[]
  setPhones: (phones: IPhone[]) => void
}