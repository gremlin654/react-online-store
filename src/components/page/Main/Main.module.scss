.main__container {
  height: 100%;
}
.filter__ {
  &container {
    grid-gap: 20px;
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    margin: 20px 0 30px;
    width: 100%;
    @media screen and (max-width: 1024px) {
      grid-template-columns: repeat(1, 1fr);
    }
  }
  &group {
    backdrop-filter: blur(3px);
    background-color: rgba(19, 20, 21, 0.3);
    border: 1px solid #fff;
    border-radius: 10px;
    padding: 10px 20px;
    z-index: 1;
  }
  &title {
    font-size: 18px;
    margin: 10px 0;
  }
  &by-shape {
    align-items: center;
    display: flex;
  }
  &by-size {
    align-items: center;
    display: flex;
  }
  &by-color {
    align-items: center;
    display: flex;
  }
  &by-favorite {
    align-items: center;
    display: flex;
    margin: 10px 0;
  }
}
.size__ {
  &list {
    list-style: none;
    margin: 0;
    padding: 0;
  }
  &button {
    background-color: transparent;
    background-position: 50%;
    background-repeat: no-repeat;
    background-size: contain;
    border: 2px solid #fff;
    border-radius: 50%;
    color: #fff;
    cursor: pointer;
    height: 40px;
    margin: 10px 7px;
    outline: 0;
    position: relative;
    width: 40px;
  }
}
.shape__ {
  &list {
    list-style: none;
    margin: 0;
    padding: 0;
  }
  &button {
    background-position: 50%;
    background-repeat: no-repeat;
    background-size: contain;
    border: 0;
    cursor: pointer;
    font-size: 9px;
    height: 40px;
    margin: 10px 7px;
    outline: 0;
    position: relative;
    width: 40px;
  }
}
.active {
  background-color: #cab67c;
}
.color__ {
  &list {
    list-style: none;
    margin: 0;
    padding: 0;
  }
  &button {
    border: 0;
    border-radius: 5px;
    cursor: pointer;
    font-size: 0;
    height: 30px;
    margin: 15px 10px;
    outline: 0;
    position: relative;
    width: 30px;
  }
  &button.active::after {
    border: solid rgb(19, 208, 32);
    border-width: 0 4px 4px 0;
    content: '';
    display: block;
    height: 10px;
    left: 9px;
    position: absolute;
    top: 5px;
    -webkit-transform: rotate(45deg);
    transform: rotate(45deg);
    width: 6px;
  }
}

.favorite__ {
  &button {
    background-color: #ded7d7;
    border: 0;
    border-radius: 5px;
    cursor: pointer;
    font-size: 0;
    height: 30px;
    margin: 15px 10px;
    outline: 0;
    position: relative;
    width: 30px;
  }
  &button.active::after {
    border: solid rgb(19, 208, 32);
    border-width: 0 4px 4px 0;
    content: '';
    display: block;
    height: 10px;
    left: 9px;
    position: absolute;
    top: 5px;
    -webkit-transform: rotate(45deg);
    transform: rotate(45deg);
    width: 6px;
  }
}

.items__ {
  &container {
    grid-gap: 20px;
    display: grid;
    grid-template-columns: repeat(5, 1fr);
    margin: auto;
    width: 100%;
    width: 96%;
    @media screen and (max-width: 1422px) {
      grid-template-columns: repeat(4, 1fr);
    }
    @media screen and (max-width: 1133px) {
      grid-template-columns: repeat(3, 1fr);
    }
    @media screen and (max-width: 785px) {
      grid-template-columns: repeat(2, 1fr);
    }
    @media screen and (max-width: 560px) {
      grid-template-columns: repeat(1, 1fr);
    }
  }
  &null_search {
    display: flex;
    justify-content: center;
    width: 100%;
  }
}
.slider__ {
  &title {
    display: flex;
    justify-content: center;
    font-size: 17px;
    margin: 20px 0 50px;
  }
  &container {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-content: center;
    align-items: center;
    gap: 15px;
    margin: 0 17px;
  }
  &count {
    display: flex;
    align-items: center;
    justify-content: center;
    -webkit-backdrop-filter: blur(3px);
    backdrop-filter: blur(3px);
    background-color: rgba(19, 20, 21, 0.3);
    border: 1px solid #278d9f;
    border-radius: 20px;
    color: #fff;
    font-family: Open Sans, sans-serif;
    font-size: 16px;
    font-weight: 400;
    height: 40px;
    width: 80px;
    margin: 10px 0;
    padding: 0 15px;
  }
}
.filter__ {
  &group {
    backdrop-filter: blur(3px);
    background-color: rgba(19, 20, 21, 0.3);
    border: 1px solid #fff;
    border-radius: 10px;
    padding: 10px 20px;
    z-index: 1;
  }
  &title {
    font-size: 18px;
    margin: 10px 0;
  }
  &reset {
    -webkit-column-gap: 20px;
    column-gap: 20px;
    display: flex;
    flex-wrap: wrap;
    margin: 30px 0 10px;
  }
}
.search__ {
  &container {
    backdrop-filter: blur(3px);
    background-color: rgba(19, 20, 21, 0.3);
    border: 1px solid #278d9f;
    border-radius: 20px;
    padding-left: 20px;
    position: relative;
  }
  &input {
    background-color: transparent;
    background-image: url(../../../assets/svg/search.svg);
    background-position: calc(100% - 20px);
    background-repeat: no-repeat;
    background-size: 20px;
    border: 0;
    color: #fff;
    font-size: 18px;
    height: 40px;
    margin-right: 0;
    width: 100%;
    &:focus-visible {
      outline: none;
    }
  }
  &input_close {
    background-color: transparent;
    background-position: calc(100% - 20px);
    background-repeat: no-repeat;
    background-size: 20px;
    border: 0;
    color: #fff;
    font-size: 18px;
    height: 40px;
    margin-right: 0;
    width: 100%;
    &:focus-visible {
      outline: none;
    }
  }
  &croos {
    background-image: url(../../../assets/img/cross.png);
    background-repeat: no-repeat;
    background-size: 20px;
    cursor: pointer;
    font-size: 0;
    height: 20px;
    position: absolute;
    right: 18px;
    top: 10px;
    width: 20px;
  }
}
.reset__button {
  backdrop-filter: blur(3px);
  background-color: rgba(19, 20, 21, 0.3);
  border: 1px solid #278d9f;
  border-radius: 20px;
  color: #fff;
  cursor: pointer;
  font-family: Open Sans, sans-serif;
  font-size: 16px;
  font-weight: 400;
  height: 40px;
  margin: 10px 0;
  padding: 0 25px;
}
.type__select {
  -webkit-box-align: center;
  align-items: center;
  background-color: rgb(255, 255, 255);
  border-color: rgb(204, 204, 204);
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: default;
  display: flex;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  justify-content: space-between;
  min-height: 38px;
  position: relative;
  transition: all 100ms ease 0s;
  box-sizing: border-box;
  outline: 0px !important;
  color: #000;
  width: 100%;
}

button:hover {
  transition: 0.3s;
  background-color: #278d9f;
}
