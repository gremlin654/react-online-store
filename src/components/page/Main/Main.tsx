import React, { useCallback, useEffect, useState } from 'react'
import { IPhone, LOCAL_STORAGE_KEYS, IPropsPhones } from '../../../types/types'
import classes from './Main.module.scss'
import { Items } from '../../common/Phones'
import Slider from 'rc-slider'
import 'rc-slider/assets/index.css'
import { MySelect } from '../../UI/select/MySelect'
import { phonesData } from '../../../data/phones'

const COLORS = ['белый', 'желтый', 'красный']
const COLORSCLASS = ['white', 'yellow', 'red']
const CAMERAS = ['1', '2', '3']
const SHAPE = ['Samsung', 'Apple', 'Xiaomi']
const INITFILTERSHAPES = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEYS.filterShape) ?? '[]')
const INITFILTERCAMERAS = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEYS.filterCameras) ?? '[]')
const INITFILTERCOLORS = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEYS.filterColor) ?? '[]')
const INITFILTERPOPULAR = JSON.parse(
  localStorage.getItem(LOCAL_STORAGE_KEYS.filterPopular) ?? 'false',
)
const INITFILTERRELEASE = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEYS.filterRelease) ?? '[]')
const INITFILTERQUANTITY = JSON.parse(
  localStorage.getItem(LOCAL_STORAGE_KEYS.filterQuantity) ?? '[]',
)
const INITSEARCH = localStorage.getItem(LOCAL_STORAGE_KEYS.search) ?? ''
const INITSORTEDBYPARAMETER = localStorage.getItem(LOCAL_STORAGE_KEYS.sortedByParameter) ?? ''

export const Main = ({ phones, setPhones }: IPropsPhones) => {
  const [filterShape, setFilterShape] = useState<string[]>(INITFILTERSHAPES)
  const [filterCameras, setFilterCameras] = useState<string[]>(INITFILTERCAMERAS)
  const [filterColor, setFilterColor] = useState<string[]>(INITFILTERCOLORS)
  const [filterPopular, setPopular] = useState<boolean>(INITFILTERPOPULAR)
  const [filterRelease, setFilterRelease] = useState<number[]>(INITFILTERRELEASE)
  const [filterQuantity, setFilterQuantity] = useState<number[]>(INITFILTERQUANTITY)
  const [search, setSearch] = useState<string>(INITSEARCH)
  const [sortedByParameters, setSortedByParameters] = useState<string>(INITSORTEDBYPARAMETER)
  const [filterClear, setFilterClear] = useState<boolean>(false)
  const [allClear, setAllClear] = useState<boolean>(false)

  const changeFilterShape = useCallback(
    (parameter: string) => {
      if (filterShape.includes(parameter)) {
        setFilterShape(filterShape.filter((shape) => shape !== parameter))
        localStorage.setItem(
          LOCAL_STORAGE_KEYS.filterShape,
          JSON.stringify(filterShape.filter((shape) => shape !== parameter)),
        )
      } else {
        setFilterShape([...filterShape, parameter])
        localStorage.setItem(
          LOCAL_STORAGE_KEYS.filterShape,
          JSON.stringify([...filterShape, parameter]),
        )
      }
    },
    [filterShape],
  )

  const changeFilterCameras = useCallback(
    (parameter: string) => {
      if (filterCameras.includes(parameter)) {
        setFilterCameras(filterCameras.filter((camera) => camera !== parameter))
        localStorage.setItem(
          LOCAL_STORAGE_KEYS.filterCameras,
          JSON.stringify(filterCameras.filter((camera) => camera !== parameter)),
        )
      } else {
        setFilterCameras([...filterCameras, parameter])
        localStorage.setItem(
          LOCAL_STORAGE_KEYS.filterCameras,
          JSON.stringify([...filterCameras, parameter]),
        )
      }
    },
    [filterCameras],
  )

  const changeFilterColor = useCallback(
    (parameter: string) => {
      if (filterColor.includes(parameter)) {
        setFilterColor(filterColor.filter((color) => color !== parameter))
        localStorage.setItem(
          LOCAL_STORAGE_KEYS.filterColor,
          JSON.stringify(filterColor.filter((color) => color !== parameter)),
        )
      } else {
        setFilterColor([...filterColor, parameter])
        localStorage.setItem(
          LOCAL_STORAGE_KEYS.filterColor,
          JSON.stringify([...filterColor, parameter]),
        )
      }
    },
    [filterColor],
  )

  const changeFilterPopular = useCallback(
    (parameter: boolean) => {
      if (filterPopular === parameter) {
        setPopular(!filterPopular)
        localStorage.setItem(LOCAL_STORAGE_KEYS.filterPopular, JSON.stringify(!filterPopular))
      } else {
        setPopular(parameter)
        localStorage.setItem(LOCAL_STORAGE_KEYS.filterPopular, JSON.stringify(parameter))
      }
    },
    [filterPopular],
  )
  const changeFilterRelease = useCallback(
    (parameter: number[] | number) => {
      setFilterRelease(parameter as number[])
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterRelease, JSON.stringify(parameter as number[]))
    },
    [filterRelease],
  )
  const changeFilterQuntity = useCallback(
    (parameter: number[] | number) => {
      setFilterQuantity(parameter as number[])
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterQuantity, JSON.stringify(parameter as number[]))
    },
    [filterQuantity],
  )

  const searchPhones = useCallback(
    (parameter: string, e: React.ChangeEvent<HTMLInputElement>) => {
      e.preventDefault()
      setSearch(parameter)
      localStorage.setItem(LOCAL_STORAGE_KEYS.search, parameter)
    },
    [search],
  )

  const resetSearchPhones = useCallback(() => {
    setSearch('')
    localStorage.setItem(LOCAL_STORAGE_KEYS.search, '')
  }, [search])

  const changeSortedByParameters = useCallback(
    (parameter: string) => {
      setSortedByParameters(parameter)
      console.log(parameter)
      localStorage.setItem(LOCAL_STORAGE_KEYS.sortedByParameter, parameter)
    },
    [sortedByParameters],
  )

  useEffect(() => {
    let filteredData: IPhone[] = phonesData
    const filterLocalColor: string[] = JSON.parse(
      localStorage.getItem(LOCAL_STORAGE_KEYS.filterColor) ?? '[]',
    )
    const filterLocalCameras: string[] = JSON.parse(
      localStorage.getItem(LOCAL_STORAGE_KEYS.filterCameras) ?? '[]',
    )
    const filterLocalShape: string[] = JSON.parse(
      localStorage.getItem(LOCAL_STORAGE_KEYS.filterShape) ?? '[]',
    )
    const filterLocalPopular: boolean = JSON.parse(
      localStorage.getItem(LOCAL_STORAGE_KEYS.filterPopular) ?? 'false',
    )
    const filterLocalRelease: number[] = JSON.parse(
      localStorage.getItem(LOCAL_STORAGE_KEYS.filterRelease) ?? '[]',
    )
    const filterLocalQuantity: number[] = JSON.parse(
      localStorage.getItem(LOCAL_STORAGE_KEYS.filterQuantity) ?? '[]',
    )
    const Search: string = localStorage.getItem(LOCAL_STORAGE_KEYS.search) ?? ''
    const sortedLocalByParameters: string =
      localStorage.getItem(LOCAL_STORAGE_KEYS.sortedByParameter) ?? ''

    if (sortedLocalByParameters === 'model') {
      filteredData = [...filteredData].sort((a, b) => a['model'].localeCompare(b['model']))
    }
    if (sortedLocalByParameters === 'modelM') {
      filteredData = [...filteredData].sort((a, b) => b['model'].localeCompare(a['model']))
    }
    if (sortedLocalByParameters === 'quantity') {
      filteredData = [...filteredData].sort((a, b) => a['quantity'] - b['quantity'])
    }
    if (sortedLocalByParameters === 'quantityM') {
      filteredData = [...filteredData].sort((a, b) => b['quantity'] - a['quantity'])
    }
    if (sortedLocalByParameters === 'release') {
      filteredData = [...filteredData].sort((a, b) => a['release'] - b['release'])
    }
    if (sortedLocalByParameters === 'releaseM') {
      filteredData = [...filteredData].sort((a, b) => b['release'] - a['release'])
    }

    if (filterLocalColor.length) {
      filteredData = filteredData.filter((post) => filterColor.includes(post.color))
    }
    if (filterLocalCameras.length) {
      filteredData = filteredData.filter((post) => filterCameras.includes(post.cameras))
    }
    if (filterLocalShape.length) {
      filteredData = filteredData.filter((post) => filterShape.includes(post.shape))
    }
    if (filterLocalPopular) {
      filteredData = filteredData.filter((post) => post.popular === true)
    }
    if (filterLocalRelease.length) {
      filteredData = filteredData.filter(
        (post) => post.release >= filterLocalRelease[0] && post.release <= filterLocalRelease[1],
      )
    }
    if (filterLocalQuantity.length) {
      filteredData = filteredData.filter(
        (post) =>
          post.quantity >= filterLocalQuantity[0] && post.quantity <= filterLocalQuantity[1],
      )
    }
    if (Search) {
      setSearch(Search)
      filteredData = filteredData.filter((post) =>
        post.model.toLowerCase().includes(Search.toLowerCase()),
      )
    }

    if (filterClear) {
      setFilterClear(false)
      setFilterColor([])
      setFilterCameras([])
      setPopular(false)
      setFilterShape([])
      setSearch('')
      setFilterRelease([2000, 2022])
      setFilterQuantity([1, 12])
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterColor, JSON.stringify([]))
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterCameras, JSON.stringify([]))
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterShape, JSON.stringify([]))
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterPopular, JSON.stringify(false))
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterRelease, JSON.stringify([2000, 2022]))
      localStorage.setItem(LOCAL_STORAGE_KEYS.filterQuantity, JSON.stringify([1, 12]))
      localStorage.setItem(LOCAL_STORAGE_KEYS.search, '')
      filteredData = phonesData
    }
    if (allClear) {
      setAllClear(false)
      setFilterColor([])
      setFilterCameras([])
      setPopular(false)
      setFilterShape([])
      setSearch('')
      setFilterRelease([2000, 2022])
      setFilterQuantity([1, 12])
      setSortedByParameters('')
      localStorage.clear()
      filteredData = phonesData
      window.location.reload()
    }
    setPhones(filteredData)
  }, [
    filterColor,
    filterCameras,
    filterShape,
    filterPopular,
    search,
    filterRelease,
    filterQuantity,
    sortedByParameters,
    allClear,
    filterClear,
  ])

  return (
    <main className={classes.main}>
      <div className={classes.main__container}>
        <div className={classes.filter__container}>
          <div className={classes.filter__group}>
            <h4 className={classes.filter__title}>Фильтры по значению</h4>
            <div>
              <div className={classes.filter__by_shape}>
                Производитель:
                <ul className={classes.shape__list}>
                  {SHAPE.map((shape: string) => (
                    <button
                      className={`${classes.shape__button} ${
                        filterShape.includes(shape) ? classes.active : ''
                      }`}
                      key={shape}
                      onClick={() => {
                        changeFilterShape(shape)
                      }}
                    >
                      {shape}
                    </button>
                  ))}
                </ul>
              </div>
              <div className={classes.filter__by_size}>
                Количество камер:
                <ul className={classes.size__list}>
                  {CAMERAS.map((camera) => (
                    <button
                      className={`${classes.size__button} ${
                        filterCameras.includes(camera) ? classes.active : ''
                      }`}
                      key={camera}
                      data-testid={camera}
                      onClick={() => {
                        changeFilterCameras(camera)
                      }}
                    >
                      {camera}
                    </button>
                  ))}
                </ul>
              </div>
              <div className={classes.filter__by_color}>
                Цвет:
                <ul className={classes.color__list}>
                  {COLORS.map((color: string, index: number) => (
                    <button
                      className={`${classes.color__button} ${COLORSCLASS[index]} ${
                        filterColor.includes(color) ? classes.active : ''
                      }`}
                      onClick={() => {
                        changeFilterColor(color)
                      }}
                      key={color}
                      data-testid={color}
                    >
                      {color}
                    </button>
                  ))}
                </ul>
              </div>
              <div className={classes.filter__by_favorite}>
                Только популярные:
                <div>
                  <button
                    className={`${classes.favorite__button} ${filterPopular ? classes.active : ''}`}
                    onClick={() => {
                      changeFilterPopular(true)
                    }}
                  ></button>
                </div>
              </div>
            </div>
          </div>
          <div className={classes.filter__group}>
            <h4 className={classes.filter__title}>Фильтры по диапазону</h4>
            <div>
              <div>
                <h4 className={classes.slider__title}>По году выхода</h4>
                <div className={classes.slider__container}>
                  <div className={classes.slider__count}>
                    {filterRelease[0] ? filterRelease[0] : 2000}
                  </div>
                  <Slider
                    range
                    step={1}
                    min={2000}
                    max={2022}
                    defaultValue={filterRelease}
                    value={filterRelease}
                    onChange={(value) => changeFilterRelease(value)}
                  />
                  <div className={classes.slider__count}>
                    {filterRelease[1] ? filterRelease[1] : 2022}
                  </div>
                </div>
              </div>
              <div>
                <h4 className={classes.slider__title}>По количеству на складе</h4>
                <div className={classes.slider__container}>
                  <div className={classes.slider__count}>
                    {filterQuantity[0] ? filterQuantity[0] : 1}
                  </div>
                  <Slider
                    range
                    step={1}
                    min={1}
                    max={10}
                    value={filterQuantity}
                    defaultValue={filterQuantity}
                    onChange={(value) => changeFilterQuntity(value)}
                  />
                  <div className={classes.slider__count}>
                    {filterQuantity[1] ? filterQuantity[1] : 10}
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div className={classes.filter__group}>
            <div className={classes.filter__search}>
              <h4 className={classes.filter__title}>Поиск</h4>
              <div className={classes.search__container}>
                <input
                  className={`${search ? classes.search__input_close : classes.search__input}`}
                  type='text'
                  placeholder='Поиск...'
                  value={search}
                  onChange={(e) => searchPhones(e.currentTarget.value, e)}
                  autoFocus
                  data-testid='input-search'
                />
                <div
                  className={`${search ? classes.search__croos : ''}`}
                  onClick={resetSearchPhones}
                ></div>
              </div>
            </div>
            <div className={classes.filter__type}>
              <h4 className={classes.filter__title}>Сортировка</h4>
              <MySelect
                value={sortedByParameters}
                onChange={(selectedSort) => changeSortedByParameters(selectedSort)}
                className={classes.type__select}
                defaultValue='По названию, от А до Я'
                options={[
                  {
                    value: 'modelM',
                    name: 'По названию, от Я до А',
                  },
                  {
                    value: 'release',
                    name: 'По году, по возростанию',
                  },
                  {
                    value: 'releaseM',
                    name: 'По году, по убыванию',
                  },
                  {
                    value: 'quantity',
                    name: 'По количеству, по возростанию',
                  },
                  {
                    value: 'quantityM',
                    name: 'По количеству, по убыванию',
                  },
                ]}
              />
            </div>
            <div className={classes.filter__reset}>
              <button
                className={classes.reset__button}
                onClick={() => {
                  setFilterClear(true)
                }}
              >
                Сброс фильтров
              </button>
              <button
                className={classes.reset__button}
                onClick={() => {
                  setAllClear(true)
                }}
              >
                Сброс настроек
              </button>
            </div>
          </div>
        </div>
        <Items className={classes.items__container} phones={phones} setPhones={setPhones} />
      </div>
    </main>
  )
}
