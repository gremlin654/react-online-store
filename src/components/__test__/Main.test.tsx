import { render, screen, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { IPhone } from '../../types/types'
import { Main } from '../page/Main/Main'
import '@testing-library/jest-dom/extend-expect'

describe('Main', () => {
  test('filters', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    const filterBlockOne = screen.getByText(/Производитель:/i)
    const filterBlockTwo = screen.getByText(/Количество камер:/i)
    const filterBlockThree = screen.getByText(/Цвет:/i)
    const filterBlockFour = screen.getByText(/Только популярные:/i)

    expect(filterBlockOne).toBeInTheDocument()
    expect(filterBlockTwo).toBeInTheDocument()
    expect(filterBlockThree).toBeInTheDocument()
    expect(filterBlockFour).toBeInTheDocument()
  })
  test('screen buttons color', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    const buttonElement = screen.getByTestId(/красный/i)
    expect(buttonElement).toMatchSnapshot()
  })
  test('input', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    expect(screen.queryByTestId('input-search')).toContainHTML('')
  })
  test('input have focus', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    const input = screen.getByPlaceholderText(/Поиск.../i)
    expect(input).toHaveFocus()
  })
  test('input event', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    const input = screen.getByPlaceholderText(/Поиск.../i)
    expect(screen.queryByTestId('input-search')).toContainHTML('')
    fireEvent.input(input, {
      target: { value: '123' },
    })
    expect(screen.queryByTestId('input-search')).toContainHTML('123')
  })
})

test('does not should element', () => {
  render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
  const filterBlock = screen.queryAllByText(/Разновидность:/i)
  expect(filterBlock).toStrictEqual([])
})

describe('should buttons styles', () => {
  test('button color have class active', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    const btnColor1 = screen.getByTestId(/белый/i)
    const btnColor2 = screen.getByTestId(/желтый/i)
    const btnColor3 = screen.getByTestId(/красный/i)
    fireEvent.click(btnColor1)
    fireEvent.click(btnColor2)
    fireEvent.click(btnColor3)
    expect(btnColor1).toHaveClass('active')
    expect(btnColor2).toHaveClass('active')
    expect(btnColor3).toHaveClass('active')
  })
  test('button color have not class active', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    const btnColor1 = screen.getByTestId(/белый/i)
    const btnColor2 = screen.getByTestId(/желтый/i)
    const btnColor3 = screen.getByTestId(/красный/i)
    expect(btnColor1).not.toHaveClass('active')
    expect(btnColor2).not.toHaveClass('active')
    expect(btnColor3).not.toHaveClass('active')
  })
  test('button hover', () => {
    render(<Main phones={[]} setPhones={(phones: IPhone[]) => {}} />)
    const btn = screen.getByTestId(/3/i)
    userEvent.hover(btn)
    expect(userEvent.hover(btn))
  })
})
