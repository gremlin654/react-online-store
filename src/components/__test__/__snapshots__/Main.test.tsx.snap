// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main elements should should buttons color 1`] = `
<button
  class="color__button red "
  data-testid="красный"
>
  красный
</button>
`;

exports[`Main elements should should screen buttons color 1`] = `
<button
  class="color__button red "
  data-testid="красный"
>
  красный
</button>
`;
